images/virtualization-artifact/pkg/eventrecord/eventrecorderlogger.go

```diff
@@ -0,0 +1,123 @@
+type EventRecorderLogger interface {
+	Event(object client.Object, eventtype, reason, message string)
+
+	// Eventf is just like Event, but with Sprintf for the message field.
+	Eventf(involved client.Object, eventtype, reason, messageFmt string, args ...interface{})
+
+	// AnnotatedEventf is just like eventf, but with annotations attached
+	AnnotatedEventf(involved client.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{})
+
+	WithLogging(logger infoLogger) EventRecorderLogger
+}
+
+// NewEventRecorderLogger implements EventRecorderLogger.
+// It creates recorder for each reason to workaround low qps limit of 1 event per 5 minutes.
+func NewEventRecorderLogger(recorderProducer recorderProducer, controllerName string) EventRecorderLogger {
+	return &EventRecorderLoggerImpl{
+		recorderProducer: recorderProducer,
+		controllerName:   controllerName,
+	}
+}
+
+// EventRecorderLoggerImpl implements Event recorder that also log event object.
+type EventRecorderLoggerImpl struct {
+	controllerName   string
+	recorderProducer recorderProducer
+	logger           infoLogger
+}
+
+func (e *EventRecorderLoggerImpl) WithLogging(logger infoLogger) EventRecorderLogger {
+	return &EventRecorderLoggerImpl{
+		controllerName:   e.controllerName,
+		recorderProducer: e.recorderProducer,
+		logger:           logger,
+	}
+}
+
+// Event logs info about event and records it as Event resource.
+func (e *EventRecorderLoggerImpl) Event(object client.Object, eventtype, reason, message string) {
+	e.logf(object, eventtype, reason, message)
+	recorder := e.recorderProducer.GetEventRecorderFor(e.recorderKey(reason))
+	recorder.Event(object, eventtype, reason, message)
+}
+
+// Eventf logs info about event and records it as Event resource. The difference with Event method is that it formats message.
+func (e *EventRecorderLoggerImpl) Eventf(object client.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+	e.logf(object, eventtype, reason, messageFmt, args...)
+	recorder := e.recorderProducer.GetEventRecorderFor(e.recorderKey(reason))
+	recorder.Eventf(object, eventtype, reason, messageFmt, args...)
+}
+
+// AnnotatedEventf logs info about event and records it as Event resource.
+// The difference with the Event method is that it formats message and add annotations to the Event resource.
+func (e *EventRecorderLoggerImpl) AnnotatedEventf(object client.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
+	e.logf(object, eventtype, reason, messageFmt, args...)
+	recorder := e.recorderProducer.GetEventRecorderFor(e.recorderKey(reason))
+	recorder.AnnotatedEventf(object, annotations, eventtype, reason, messageFmt, args...)
+}
+
+func (e *EventRecorderLoggerImpl) logf(involved client.Object, eventtype, reason, messageFmt string, args ...interface{}) {
+	if e.logger == nil {
+		return
+	}
+	e.logger.Info(
+		fmt.Sprintf(messageFmt, args...),
```

Comment:
Can we have a problem with this? If the fmt string and args do not match.

Reply:
At least Goland can catch these problems. Didn't check golangci-lint though.
<img width="595" alt="image" src="https://github.com/user-attachments/assets/b1d7cc08-afcf-4d17-afba-16d90acb3982" />